pymc3/distributions/__init__.py

```diff
@@ -11,8 +11,125 @@
 #   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
+from functools import singledispatch
+from typing import Optional
 
-from pymc3.distributions import shape_utils, timeseries, transforms
+import theano.tensor as tt
+
+from theano.tensor.random.op import Observed
+from theano.tensor.var import TensorVariable
+
+
+def get_rv_var_and_value(
+    rv_var: TensorVariable,
+    rv_value: Optional[TensorVariable] = None,
+) -> TensorVariable:
+
+    if rv_value is None:
+        if rv_var.owner and isinstance(rv_var.owner.op, Observed):
+            rv_var, rv_value = rv_var.owner.inputs
+        elif hasattr(rv_var.tag, "value_var"):
+            rv_value = rv_var.tag.value_var
+        else:
+            raise ValueError("value is unspecified")
+
+    return rv_var, rv_value
+
+
+def logpt(
+    rv_var: TensorVariable,
+    rv_value: Optional[TensorVariable] = None,
+    jacobian: bool = True,
+    scaling: Optional[TensorVariable] = None,
+    **kwargs
+) -> TensorVariable:
+    """Get a graph of the log-likelihood for a random variable at a point."""
+
+    rv_var, rv_value = get_rv_var_and_value(rv_var, rv_value)
+    rv_node = rv_var.owner
+
+    if not rv_node:
+        raise TypeError("rv_var must be the output of a RandomVariable Op")
+
+    rng, size, dtype, *dist_params = rv_node.inputs
+
+    if jacobian:
+        logp_var = _logp(rv_node.op, rv_value, *dist_params, **kwargs)
+    else:
+        logp_var = _logp_nojac(rv_node.op, rv_value, *dist_params, **kwargs)
+
+    if scaling:
+        logp_var *= scaling
+
+    if rv_var.name is not None:
+        logp_var.name = "__logp_%s" % rv_var.name
+
+    return logp_var
+
+
+@singledispatch
+def _logp(op, value, *dist_params, **kwargs):
+    return tt.zeros_like(value)
+
+
+def logcdf(rv_var, rv_value, **kwargs):
+
+    rv_var, rv_value = get_rv_var_and_value(rv_var, rv_value)
+    rv_node = rv_var.owner
+
+    if not rv_node:
+        raise TypeError()
+
+    rng, size, dtype, *dist_params = rv_node.inputs
+
+    return _logcdf(rv_node.op, rv_value, *dist_params, **kwargs)
+
+
+@singledispatch
+def _logcdf(op, value, *args, **kwargs):
```

Comment:
I wonder if now would also be a good time to add lccdf (complementary lcdf) methods? Sometimes the lccdf is more efficient than the lcdf.

Reply:
That's out of scope for this PR/discussion.

Reply:
I didn't mean to implement them but to have a placeholder not implemented. But you are right it doesn't need to be here